### DISCUSSION TITLE
Handling of timezone in timestamptz dumping

### DISCUSSION THREAD

**Discussion by dvarrazzo** [psycopg/psycopg#56] [Ideas]
For PostgreSQL, timestmaptz is not, as the tin says, a timestamp with a timezone, but [an UTC timestamp](https://www.postgresql.org/docs/current/datatype-datetime.html) which, at time time of output, is presented in the timezone configured in the [`TimeZone` setting](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-TIMEZONE). The format is further controlled by the [`DateStyle` setting](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-DATESTYLE), which is actually two independent settings rolled into one.

The output style, in the most usable setting, is of a form like `2021-05-14 10:51:27.164228+01` when dumped in text format. However, when dumped in the binary format, it is just the number of µs from 2000-01-01 00:00:00 UTC in an int64. The `TimeZone` setting doesn't affect it.

psycopg2 only dumps the text format and returns a dateime with a fixed offset timezone as parsed from the string.

psycopg3 supports both text and binary format. The first dumper written was a porting of the psycopg2 text dumper. The natural way of writing a binary dumper makes it produce UTC timezones regardless of the `timezone` setting. The inconsistency between the text and the binary format is pretty bad and I think it should be avoided.

```python
>>> print(cnn.cursor(binary=True).execute("select now()").fetchone()[0])
2021-05-14 10:40:15.046149+00:00
>>> print(cnn.cursor(binary=False).execute("select now()").fetchone()[0])
2021-05-14 11:40:15.046149+01:00
```

the choices are either to make the text dumper return always utc or to make the binary dumper return the connection timezone. The first is trivial (and current implementation in the `binary-datetime` branch as of 4326008). However it might be a bad idea to change behaviour from psycopg2 so radically (as [@asqui maintains](https://github.com/psycopg/psycopg2/issues/1272#issuecomment-840858721)).

Doing the opposite, returning datetimes in connection encoding, has a slight performance penalty, requires us to depend on the `pytz` library, and doesn't cover all the cases (pytz doesn't parse [everything Postgres can emit](https://www.postgresql.org/docs/13/datetime-posix-timezone-specs.html)). But I am inclined to think it would be a more user-friendly choice.

Also note that `pytz` doesn't support seconds in the timezone. This was a limitation of Python before 3.7 (seee https://github.com/psycopg/psycopg2/issues/1272 and [psycopg2 handliing](https://www.psycopg.org/docs/usage.html?highlight=timezone#time-zones-handling)). This can be probably fixed by moving the error to the time offset rather than in the timestamp. For instance:

```
piro=# set timezone to 'Europe/Paris';
SET
piro=# select '1789-07-14Z'::timestamptz;
┌──────────────────────────────┐
│         timestamptz          │
├──────────────────────────────┤
│ 1789-07-14 00:09:21+00:09:21 │
└──────────────────────────────┘
```

Even if pytz and Python pre-3.7 cannot represent the 21s in the timezone, getting to the UTC timezone and then applying the timezone in Python results in the correct time.

```python
>>> cnn.execute("set timezone to 'Europe/Paris'");

# In the current psycopg3 implementation always returning UTC
>>> print(cnn.execute("select '1789-07-14Z'::timestamptz").fetchone()[0])
1789-07-14 00:00:00+00:00

# This is what we might do instead and, unlike psycopg2 as of 2.8.6, it is correct.
>>> tz = pytz.timezone("Europe/Paris")
>>> print(cnn.execute("select '1789-07-14Z'::timestamptz").fetchone()[0].astimezone(tz))
1789-07-14 00:09:00+00:09
```

**Comment by olasd** [comment]
(Sorry, small rant ahead!)

In general, I think relying on the timezone definitions of the PostgreSQL server when deserializing `timestamptz` columns is fraught with peril. What the client app thinks is the `Europe/Paris` timezone could very well be different than what the server thinks it is. Even though that specific timezone definition is probably not changing any time soon (although, maybe we're getting rid of DST in the next few years), there's a bunch of more volatile timezone definitions out there, and it's fairly easy to end up in a state where the client and the server have mismatched expectations for what a textual timezone definition means.

I don't think using pytz would be necessary in general, as PostgreSQL will use its baked-in Olsen database to generate the relevant timezone offset for the actual point in time of the timeztamptz value returned, but the data in transit (in the text protocol) will always be a *static* UTC offset in seconds. PostgreSQL won't be able to dump you the full timezone definition -- e.g. "Europe/Paris", or, probably better, all the offset and DST rules that correspond to what's currently compiled in as the "Europe/Paris" definition. The psycopg2 `FixedOffsetTimezone` object (with added support for seconds) should be sufficient to handle all the values that PostgreSQL will return.

In general I think that, rather than encouraging the fairly common confusion that `timestamptz` columns *store* a timezone, a PostgreSQL library should strongly encourage the use of (inambiguous) UTC points-in-time (that is, what the binary protocol returns), return such objects by default, and leave the client app to handle timezone conversions explicitly. This would also encourage client apps to explicitly store timezone information if they need it (*alongside* the inambiguous point-in-time timestamptz columns).

As a convenience for users currently relying on the connection/server timestamp setting, the library could provide a way for client apps to set an explicit timezone object which would be used to "convert" (well, it would be set as an attribute on) all the `datetime` objects generated by the library. You could even set that to default to the current local timezone definition as generated by pytz. Using an explicit timezone object controlled by the client app means that you can have "timezone" support for all timestamptz values, regardless of whether the text or binary protocol is being used. You could also have a flag value which would generate the relevant `FixedOffsetTimezone` if the cursor uses the text protocol (but, again, as an opt-in feature).

**Reply by dvarrazzo** [reply]
I thought about whether the difference in the timezone database would be a problem. I think that making sure that an eventual inconsistency between client and server tzinfo data results in a discrepancy in the timezone offset rather than in the timestamp would be good enough: it is better than what psycopg2 does at the moment (an error of up to 30s in the timestamp).

We can definitely offer different objects to customise the transformation from timestamptz to datetime:

1) always utc,
2) connection timezone,
3) fixed offset as per postgres result (for text only), even
4) fixed offset or
5) client time 

which I don't think anybody has ever asked. The questions are: which one should be the default, and is there a compelling use case to offer alternatives.

The option 1) of returning always utc is appealing for me, but the confusion that would arise comparing with psql results, and the expectation porting from psycopg2 are factors worth considering and moving the choice to 2) connection timezone.

**Comment by dvarrazzo** [comment]
Looking better into what's available, the stdlib [zoneinfo](https://docs.python.org/3/library/zoneinfo.html) pacakge and its backport seem a better choice than the third party `pytz`.

**Comment by dvarrazzo** [comment]
The branch `tzinfo` has a first cut of the feature: this is [how it looks](https://github.com/psycopg/psycopg3/compare/binary-datetime...tzinfo?expand=1) compared to the fixed UTC implementation.

**Comment by dvarrazzo** [comment]
In order to answer to #223 I ended up writing a long description of what's going on with timezone. It turns out that the problem in #223 can be probably easily solved by installing the `tzdata` package.

I think the description is interesting but the OP deserves a less rambly answer, so I'm moving that description here and give them the short version.

----

A recap:

- PostgreSQL can send data in two formats: text or binary. Psycopg can handle both. It is desirable that a different format doesn't change the value returned.
- PostgreSQL doesn't really handle "timestamp with timezone" as a data type. It's an illusion. What handles is only timestamps without timezone, intended as UTC timestamps, which, in text mode, are converted to the client timezone - the value you can see running `SHOW timezone`. You can't store the datum "the first of January in Moscow time" if the client timezone is in Rome.
```
piro=# set timezone to 'Europe/Rome';
SET
piro=# select '2022-01-01 00:00:00+03'::timestamptz;
┌────────────────────────┐
│      timestamptz       │
├────────────────────────┤
│ 2021-12-31 22:00:00+01 │
└────────────────────────┘
```

what you get is an equivalent timestamp in the client timezone.

- In binary format, we don't get offset info from the server, we only get the microseconds from an epoch in UTC; the client timezone setting doesn't change it.
- So, if we want to return the same value, in both text and binary format, we need to do some gymnastic and use the extra info of the timezone given by the connection:
  - In text mode we shift the value by the offset, so we have an UTC, and then convert it to the client timezone, on the client.
  - In binary mode we already have an UTC, so we just to convert it as the second step above.
- The connection timezone is given as a name, like `Europe/Rome`. The [ZoneInfo object](https://docs.python.org/3/library/zoneinfo.html#zoneinfo.ZoneInfo) knows about them and can tell the offset and DST change rules, using the underlying OS [Timezone database](https://en.wikipedia.org/wiki/Tz_database).
- Which Windows doesn't have, so you get an UTC timezone.

---

All this explains why things are like this. psycopg2 does things in a more naive way. Because it only handles the text mode it only cares about parsing the text value and uses the offset like a fixed offset. Unfortunately for Postgres that's not a fixed offset: it's the offset that the client timezone has _at that time_. If you start doing dates arithmetics you can get wrong results.

```
In [1]: import psycopg2                                                                                                             
In [2]: cnn = psycopg2.connect("")                                                                                                  
In [3]: cur = cnn.cursor()                                                                                                          
In [6]: cur.execute("set timezone to 'europe/rome'")                                                                                
In [9]: cur.execute("select '2022-01-01T00:00'::timestamptz")                                                                       
In [10]: cur.fetchone()[0]                                                                                                          
Out[10]: datetime.datetime(2022, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))

In [11]: import datetime as dt                                                                                                      
In [13]: _10 + dt.timedelta(days=180)                                                                                               
Out[13]: datetime.datetime(2022, 6, 30, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))
```

This timestamp is wrong: the offset in Rome in summer is +2 hours. If we got there from postgres:

```
In [14]: cur.execute("select '2022-01-01T00:00'::timestamptz + '180 days'::interval")                                               
In [15]: cur.fetchone()[0]                                                                                                          
Out[15]: datetime.datetime(2022, 6, 30, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))

In [16]: _13 == _15                                                                                                                 
Out[16]: False
```

Using the trivial way of doing it, Python and Postgres disagree on timestamps. I think the way psycopg 3 does is an improvement:

```
In [1]: import psycopg
In [2]: cnn = psycopg.connect("")
In [4]: cnn.execute("set timezone to 'europe/rome'")
In [5]: cnn.rollback()
In [6]: cnn.execute("set timezone to 'europe/rome'")
In [7]: cnn.execute("select '2022-01-01T00:00'::timestamptz").fetchone()[0]
Out[7]: datetime.datetime(2022, 1, 1, 0, 0, tzinfo=backports.zoneinfo.ZoneInfo(key='Europe/Rome'))

In [8]: import datetime as dt
In [9]: _7 + dt.timedelta(days=180)
Out[9]: datetime.datetime(2022, 6, 30, 0, 0, tzinfo=backports.zoneinfo.ZoneInfo(key='Europe/Rome'))

In [10]: cnn.execute("select '2022-01-01T00:00'::timestamptz + '180 days'::interval").fetchone()[0]
Out[10]: datetime.datetime(2022, 6, 30, 0, 0, tzinfo=backports.zoneinfo.ZoneInfo(key='Europe/Rome'))

In [11]: _9 == _10
Out[11]: True
```

The timezone database is available by default on Linux and macOS. With testing I discovered that it isn't on Windows. So the timezones returned are correct (they identify the correct utc timestamp) but are in UTC timezone rather than the client timezone - so correct but not necessarily what the client expects.

**Reply by sharmankin** [reply]
Thanks for the detailed explanation
